### PR TITLE
Feature/kitosudv 2114 task 2160

### PR DESCRIPTION
**Pull request by JacobMalling**
## KITOS Pull request template
The following procedure dictates the steps needed before a Pull request can be merged into master.

- [x] **Implement**: 
      *All requirements are implemented and unit tests are green*
      
- [x] **Merge master into branch / rebase with master**: 
      *Make sure you are testing your changes and how they co-exist with the latest version of master*
      
- [x] **Green on integration**: 
      *All integration tests are green on integration*
      
- [x] **Request review**: 
      *Tag whomever you wish to review your code*
      
- [x] **Review completed**: 
      *Reviewer ticks this box when review comments have been submitted*
      
- [x] **Changes**: 
      *PR owner and reviewer agrees on which changes must be made and the changes are committed.*
      
- [x] **Merge master into branch / rebase with master**: 
      *Make sure you are testing your changes and how they co-exist with the latest version of master*
      
- [x] **Green on integration**: 
      *All integration tests are green on integration*

### CODE REVIEW

File: Core.ApplicationServices/GDPR/Write/DataProcessingRegistrationWriteService.cs [hunk 1]
```diff
@@ -109,7 +109,50 @@ public class DataProcessingRegistrationWriteService : IDataProcessingRegistratio
...
+        {
+            return dpr.WithOptionalUpdate(usageRoles.UserRolePairs, UpdateRoles);
+        }
+
+        private Result<DataProcessingRegistration, OperationError> UpdateRoles(DataProcessingRegistration dpr, Maybe<IEnumerable<UserRolePair>> userRolePairs)
+        {
+            var existingRightsList = dpr.Rights.Select(x => new UserRolePair { RoleUuid = x.Role.Uuid, UserUuid = x.User.Uuid }).ToList();
+            var newRightsList = userRolePairs.GetValueOrFallback(new List<UserRolePair>()).ToList();
+
+            foreach (var (delta, item) in existingRightsList.ComputeDelta(newRightsList, x => x))
```
Comment (mrjsawdk) [inline Core.ApplicationServices/GDPR/Write/DataProcessingRegistrationWriteService.cs:126]: ComputeDelta hader duplicates (så dør den) så du skal sikre dig at der ikke er duplicates i newRightsList inden den kaldes.

Comment (mrjsawdk) [reply Core.ApplicationServices/GDPR/Write/DataProcessingRegistrationWriteService.cs:126]: og det er jo også en del af kontrakten så hvis der er det, så er der jo badInput

File: Presentation.Web/Controllers/API/V2/External/DataProcessingRegistrations/Mapping/DataProcessingRegistrationWriteModelMapper.cs [hunk 2]
```diff
@@ -71,5 +75,21 @@ public UpdatedDataProcessingRegistrationOversightDataParameters MapOversight(Dat
...
         }
+
+        public UpdatedDataProcessingRegistrationRoles MapRoles(IEnumerable<RoleAssignmentRequestDTO> roles)
+        {
+            var roleAssignmentResponseDtos = roles.ToList();
+
+            return new UpdatedDataProcessingRegistrationRoles
+            {
+                UserRolePairs = (roleAssignmentResponseDtos.Any() ?
+                    Maybe<IEnumerable<UserRolePair>>.Some(roleAssignmentResponseDtos.Select(x => new UserRolePair
```
Comment (mrjsawdk) [inline Presentation.Web/Controllers/API/V2/External/DataProcessingRegistrations/Mapping/DataProcessingRegistrationWriteModelMapper.cs:86]: Du behøves ikke "some" factory metoden.
Du kan omskrive til
```
roleAssignmentResponseDtos.Any() ?
               roleAssignmentResponseDtos.Select(x => new UserRolePair
                    {
                        RoleUuid = x.RoleUuid,
                        UserUuid = x.UserUuid
                    }) :
                    Maybe<IEnumerable<UserRolePair>>.None
```

File: Tests.Integration.Presentation.Web/GDPR/V2/DataProcessingRegistrationApiV2Test.cs [hunk 3]
```diff
@@ -825,7 +828,84 @@ public async Task Can_PUT_With_OversightData()
...
-        private static void AssertEmptiedOversight(DataProcessingRegistrationOversightResponseDTO actual)
+        [Fact]
+        public async Task Can_POST_With_Roles()
+        {
+            //Arrange
+            var (token, user, organization) = await CreatePrerequisitesAsync();
+
+            var user1 = await CreateUser(organization);
+            var user2 = await CreateUser(organization);
+            var role = DatabaseAccess.MapFromEntitySet<DataProcessingRegistrationRole, DataProcessingRegistrationRole>(x => x.AsQueryable().First());
```
Comment (mrjsawdk) [inline Tests.Integration.Presentation.Web/GDPR/V2/DataProcessingRegistrationApiV2Test.cs:839]: Kan du ikke bare hente rollerne fra API V2 via helperen? Så sparer vi tiden på en ekstra connection til db

File: Tests.Integration.Presentation.Web/GDPR/V2/DataProcessingRegistrationApiV2Test.cs [hunk 4]
```diff
@@ -825,7 +828,84 @@ public async Task Can_PUT_With_OversightData()
...
+            var roles = new List<RoleAssignmentRequestDTO>
+            {
+                new()
+                {
+                    RoleUuid = role.Uuid,
+                    UserUuid = user1.Uuid
+                },
+                new()
+                {
+                    RoleUuid = role.Uuid,
```
Comment (mrjsawdk) [inline Tests.Integration.Presentation.Web/GDPR/V2/DataProcessingRegistrationApiV2Test.cs:849]: har vi kun 1 rolle? Vi bør måske have en ekstra ind i seed/db task (dem i kodeprojektet) så vi er sikre på at have mere end 1 til de her tests

**Review by mrjsawdk — APPROVED** [review]
fint med småting

File: Tests.Integration.Presentation.Web/GDPR/V2/DataProcessingRegistrationApiV2Test.cs (hunk 4 above)
Comment (JacobMalling) [reply Tests.Integration.Presentation.Web/GDPR/V2/DataProcessingRegistrationApiV2Test.cs:849]: Der er allerede 2. Fik bare ikke brugt nummer 2

**Pull request merged**